src/coreclr/vm/nativeimage.cpp

```diff
@@ -141,14 +141,17 @@ NativeImage *NativeImage::Open(
     LPWSTR searchPathsConfig;
     IfFailThrow(CLRConfig::GetConfigValue(CLRConfig::INTERNAL_NativeImageSearchPaths, &searchPathsConfig));
 
-    NewHolder<PEImageLayout> peLoadedImage;
+    PEImageLayoutHolder peLoadedImage;
 
     BundleFileLocation bundleFileLocation = Bundle::ProbeAppBundle(fullPath, /*pathIsBundleRelative */ true);
     if (bundleFileLocation.IsValid())
     {
-        PEImageHolder pImage = PEImage::OpenImage(fullPath, MDInternalImport_Default, bundleFileLocation);
-        peLoadedImage = pImage->GetOrCreateLayout(PEImageLayout::LAYOUT_MAPPED);
-        peLoadedImage.SuppressRelease();
+        // No need to use cache for this PE image. It is transient.
+        // We only need it to obtain the native image, which AppDomain will keep.
+        PEImageHolder pImage = PEImage::OpenImage(fullPath, MDInternalImport_NoCache, bundleFileLocation);
+        PEImageLayout* mapped = pImage->GetOrCreateLayout(PEImageLayout::LAYOUT_MAPPED);
+        mapped->AddRef();
+        peLoadedImage = mapped;
```

Comment:
My reading of the issue is that we were using `NewHolder` instead of `PEImageLayoutHolder`/`ReleaseHolder`, which meant that `SuppressRelease` wouldn't prevent the `delete` when the `NewHolder` went out of scope.

We could have switched to `PEImageLayoutHolder` and kept what this was doing before (using `SuppressRelease`). However, with the switch to calling `PEImage::OpenImage` with `MDInternalImport_NoCache`, the `pImage` will not be cached, so it (and therefore the layout created through `GetOrCreateLayout`) will be released once it goes out of scope - hence we need to do the `AddRef` instead so that we keep the `PEImageLayout` around without the `PEImage`.

Is that correct?

Reply:
If so, it might be worth having an explicit comment around that. My initial reaction looking at this was that it would make more sense to keep with the `SuppressRelease` (instead of the explicit `AddRef` and let the holder release) - before I saw the implications of the switch to `MDInternalImport_NoCache`.

Reply:
> We could have switched to PEImageLayoutHolder and kept what this was doing before (using SuppressRelease).

Yes, that would work too. `SuppressRelease` would effectively "leak" the PE image and that would keep the layout alive. We would not have a way to free the layout, but in this case it would be ok. 

On the other hand we just need this PE image to get the layout and we will not be doing that again, since AppDomain caches the native image. I figured I'd rather take the ownership of the layout (by doing AddRef) and let PE image instance be freed.

Using cache is not required here since this PE is not a part of anyone's identity. 
Since we will open the image only once anyways, I just thought we should not involve the cache, regardless if we keep/leak PE image or not.

Reply:
I will add a better comment about the whole deal.

Reply:
Added comments.